diff --git a/js/js_confirm_pop_up.js b/js/js_confirm_pop_up.js
index 014af6e..2df584d 100755
--- a/js/js_confirm_pop_up.js
+++ b/js/js_confirm_pop_up.js
@@ -26,6 +26,9 @@
                     }
                 });
             });
+            $('.js-confirm-pop-up input[type="file"]').change(function () {
+                edit = true;
+            });
 
             // Let all form submit buttons through.
             $(".js-confirm-pop-up input[type='submit']").each(function () {
